apps/builder/app/routes/oauth.ws.authorize.tsx

```diff
@@ -0,0 +1,229 @@
+  response_type: z.literal("code"),
+  redirect_uri: z.string().url(),
+
+  client_id: z.string(),
+  state: z.string(),
+  scope: z
+    .string()
+    .refine((str) => str.startsWith("project:"), {
+      message: "Only 'project:' scopes are allowed",
+    })
+    .transform((scope) => ({
+      projectId: scope.split(":")[1],
+    })),
+  code_challenge: z.string(),
+  code_challenge_method: z.literal("S256"),
+});
+
+const OAuthRedirectUriSchema = z.object({
+  redirect_uri: z.string().url(),
+});
+
+/**
+ * Based RFC 6749 and RFC 7636
+ *
+ * https://datatracker.ietf.org/doc/html/rfc6749
+ *
+ * https://datatracker.ietf.org/doc/html/rfc7636
+ */
+export const loader: LoaderFunction = async ({ request }) => {
+  try {
+    debug("Authorize request received", request.url);
+
+    const url = new URL(request.url);
+    const searchParams = Object.fromEntries(url.searchParams);
+
+    const parsedRedirect = OAuthRedirectUriSchema.safeParse(searchParams);
+
+    if (!parsedRedirect.success) {
+      debug("redirect_uri not provided in query params");
+
+      return json(
+        {
+          error: "invalid_request",
+          error_description: "No redirect_uri provided",
+          error_uri: "https://tools.ietf.org/html/rfc6749#section-3.1.2",
+        },
+        { status: 400 }
+      );
+    }
+
+    // eslint-disable-next-line camelcase
+    const { redirect_uri } = parsedRedirect.data;
+
+    // Validate the redirect_uri
+    // It is not pre-registered but it must match the AuthorizationServerOrigin
+    if (
+      getAuthorizationServerOrigin(request.url) !==
+      getAuthorizationServerOrigin(redirect_uri)
+    ) {
+      debug("redirect_uri does not match the registered redirect URIs");
+
+      return json(
+        {
+          error: "invalid_request",
+          error_description:
+            "The redirect_uri provided does not match the registered redirect URIs.",
+          error_uri: "https://tools.ietf.org/html/rfc6749#section-3.1.2",
+        },
+        { status: 400 }
+      );
+    }
+
+    let oauthError = createOauthError(redirect_uri, searchParams.state);
+
+    const parsedOAuthParams = OAuthParamsSchema.safeParse(searchParams);
+
+    if (!parsedOAuthParams.success) {
+      debug(fromError(parsedOAuthParams.error).toString());
+
+      return oauthError(
+        "invalid_request",
+        fromError(parsedOAuthParams.error).toString()
+      );
+    }
+
+    // Reinit with parsed state
+    oauthError = createOauthError(redirect_uri, parsedOAuthParams.data.state);
+
+    // client_id: Verify that the client_id is valid and corresponds to a registered client.
+    if (parsedOAuthParams.data.client_id !== env.AUTH_WS_CLIENT_ID) {
+      debug("Client is not registered");
+
+      return oauthError("unauthorized_client", "Client is not registered");
+    }
+
+    const oAuthParams = parsedOAuthParams.data;
+
+    const user = await authenticator.isAuthenticated(request);
+
+    if (user) {
```

Comment:
as an idea we could split this in a few functions, e.g. handleWithUser, handleWithoutUser 
this would make try/catch much smaller and still readable generally

Reply:
It’s linear code just a lot of linear logic IMO adding functions for this makes it less readable

Reply:
Simplified removed another if